pyat/at/physics/rdt.py

```diff
@@ -0,0 +1,575 @@
+        rdts["h10110"] = sum(-b3lm * rbetaxm * betaym / 4 * pxm * pf(1, 0))
+        rdts["h10020"] = sum(
+            -b3lm * rbetaxm * betaym / 8 * pxm * np.conj(pym * pym) * pf(1, -2)
+        )
+        rdts["h10200"] = sum(-b3lm * rbetaxm * betaym / 8 * pxm * pym * pym * pf(1, 2))
+
+    if (RDTType.TUNESHIFT in rdttype) or (RDTType.ALL in rdttype):
+        b4lm = b4l[mask_b4l]
+        betaxm = betax[mask_b4l]
+        betaym = betay[mask_b4l]
+        rdts["dnux_dJx"] = sum(3 * b4lm * betaxm * betaxm / (8 * np.pi) * nperiods)
+        rdts["dnux_dJy"] = sum(-3 * b4lm * betaxm * betaym / (4 * np.pi) * nperiods)
+        rdts["dnuy_dJy"] = sum(3 * b4lm * betaym * betaym / (8 * np.pi) * nperiods)
+
+    if (RDTType.GEOMETRIC2 in rdttype) or (RDTType.ALL in rdttype):
+        b4lm = b4l[mask_b4l]
+        betaxm = betax[mask_b4l]
+        betaym = betay[mask_b4l]
+        pxm = px[mask_b4l]
+        pym = py[mask_b4l]
+        rdts["h22000"] = sum(3 * b4lm * betaxm * betaxm / 32 * nperiods)
+        rdts["h11110"] = sum(-3 * b4lm * betaxm * betaym / 8 * nperiods)
+        rdts["h00220"] = sum(3 * b4lm * betaym * betaym / 32 * nperiods)
+        rdts["h31000"] = sum(b4lm * betaxm * betaxm / 16 * pxm * pxm * pf(2, 0))
+        rdts["h40000"] = sum(
+            b4lm * betaxm * betaxm / 64 * pxm * pxm * pxm * pxm * pf(4, 0)
+        )
+        rdts["h20110"] = sum(-3 * b4lm * betaxm * betaym / 16 * pxm * pxm * pf(2, 0))
+        rdts["h11200"] = sum(-3 * b4lm * betaxm * betaym / 16 * pym * pym * pf(0, 2))
+        # fmt: off
+        rdts["h20020"] = sum(
+            -3 * b4lm * betaxm * betaym / 32 * pxm * pxm
+            * np.conj(pym * pym) * pf(2, -2)
+        )
+        # fmt: on
+        rdts["h20200"] = sum(
+            -3 * b4lm * betaxm * betaym / 32 * pxm * pxm * pym * pym * pf(2, 2)
+        )
+        rdts["h00310"] = sum(b4lm * betaym * betaym / 16 * pym * pym * pf(0, 2))
+        rdts["h00400"] = sum(
+            b4lm * betaym * betaym / 64 * pym * pym * pym * pym * pf(0, 4)
+        )
+
+    if second_order:
+        assert nperiods == 1, "Second order available only for nperiods=1"
+
+        if (RDTType.GEOMETRIC2 in rdttype) or (RDTType.ALL in rdttype):
+            nelem = sum(mask_b3l)
+            sm = s[mask_b3l]
+            b3lm = b3l[mask_b3l]
+            betaxm = betax[mask_b3l]
+            betaym = betay[mask_b3l]
+            rbetaxm = rbetax[mask_b3l]
+            rbbetaxm = betaxm * rbetaxm
+            phixm = phix[mask_b3l]
+            phiym = phiy[mask_b3l]
+            pxm = px[mask_b3l]
+            pxm2 = pxm * pxm
+            pxm3 = pxm2 * pxm
+            cpxm = np.conj(pxm)
+            cpxm3 = np.conj(pxm3)
+            pym = py[mask_b3l]
+            pym2 = pym * pym
+            cpym2 = np.conj(pym2)
+            cpxym2 = np.conj(pxm * pym2)
+            # fmt: off
+            bsign = np.array([np.sign(sm[i] - sm) * 1j * b3lm[i] * b3lm
+                              for i in range(nelem)])
```

Comment:
This sign terms are reversed. It should be positive for all the jth elements anywhere downstream wrt to the ith element.
`np.sign(sm-sm[i])`

Reply:
@oscarxblanco , now I am confused. For me the absolute sign does not matter but if I change it in here I also have to change it for the first order terms otherwise the result will be wrong.

More generally don't we want to keep the same as in matlab? Already the coupling terms will differ from one interface to the other...

Reply:
Dear @swhite2401 ,
the sign matters because you are adding the components coming from octupoles, otherwise they could be relative.

Unfortunately, the derivations are giving me the sum term as it is written in Wang's article, and it is different in the code. I will later attach the notes of this derivation in a pdf. They are too long to be written down in markup confortably.

I understand if you prefer to keep it as in matlab, I believe nobody has complained, I guess it has been checked before.

Reply:
Also, the first order terms do not change.
The change is in the crossed-terms from sextupoles.

The error comes when implementing the summation in Eq. (46) on Wang's article. It is positive for $j>i$ and negative for $i > j$.

Reply:
Sure, what I mean is that this is more than a sign convention as opposed to a change of the sign of the total sum. If signs are not consistent between first and second orders this is a bug that needs to be corrected everywhere (but also reported to the ELEGANT developers?)

Reply:
I put the question in the elegant forum.
https://www3.aps.anl.gov/forums/elegant/viewtopic.php?f=17&t=1402

Reply:
Ah great thanks a lot! Sorry I haven't had time to come back to this lately.... meanwhile should we make a commit that gives the same results as matlab?

Reply:
Dear @swhite2401 ,
Michael Borland has answered today, he confronts the results in Elegant with OPA https://ados.web.psi.ch/opa/ . The sign that matches OPA and elegant is the same you wrote, but opposite to Chunxi's Equation.

I propose the following:
1. Document in https://atcollab.github.io/at/p/index.html all the new formulas  you are adding for the linear terms that are not in any of the citations. Mention them in the DocString.
2. Mention in  https://atcollab.github.io/at/p/index.html the difference in sign between Bengtsson's article and AT. That will clarify why first order terms are written as such.
3. Provide separately the terms produced by octupoles from the second order terms by sextupoles, i.e. do not add them. The user will add or substract them based on his own knowledge.
4. Document in  https://atcollab.github.io/at/p/index.html the difference between your summation and Chun'xi's.

Michael's response does not confirms nor denies this is an error, it is based on what another code does,  therefore, I could not approve this PR without item 3.

Reply:
Dear @oscarxblanco , I would prefer to document that we have set the signs to follow the convention used in ELEGANT for all terms and that there is descrepencies w.r.t. to the reference, such that results are identical in matlab and python.

I will provide first and second order terms separately as suggested. 

In addition, I will verify that the detuning coefficient are consistent with tracking, this should clear any ambiguity on the sign.

For what concerns RDTs it should be possible to reconstruct them using tbt tracking data but this is not so easy... I don't really have time to do that now. We can keep it for later to confirm all the signs.

Would that be acceptable for you?

Reply:
> Dear @oscarxblanco , I would prefer to document that we have set the signs to follow the convention used in ELEGANT for all terms and that there is descrepencies w.r.t. to the reference, such that results are identical in matlab and python.

OK.
> 
> I will provide first and second order terms separately as suggested.

Perfect, thank you.
> 
> In addition, I will verify that the detuning coefficient are consistent with tracking, this should clear any ambiguity on the sign.

If you refer to compare the output of `at.detuning` wrt to the terms `dnux_dJx`, `dnuy_dJy` and `dnux_dJy` this is already done, the values are correct but lead to issue #836 asking to fix the documentation only.
In addition, Bengtsson's Eq (119) depends on the absolute value of the phase advance between sextupoles, so different to the Poisson bracket the detuning has no difference for sextupoles $i,j$ or $j,i$. And the equation itself has a different notation using $j,k$ as indexes which are not explained and refer to another public document which has a different notation.
I think one would need to derive the expression for the detuning following Bengtsson definitions.


> 
> For what concerns RDTs it should be possible to reconstruct them using tbt tracking data but this is not so easy... I don't really have time to do that now. We can keep it for later to confirm all the signs.

Great, that would solve it !
> 
> Would that be acceptable for you?

Yes, of course. Thank you.